test/integration/test_tracing.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Learn more about testing at: https://juju.is/docs/sdk/testing
+from __future__ import annotations
+
+import json
+import logging
+import time
+from typing import Any, Callable
+
+import httpx
+import jubilant
+
+
+def test_direct_connection(build_charm: Callable[[], str], juju: jubilant.Juju):
+    charm_path = build_charm()
+    juju.deploy(charm_path)
+    juju.integrate('tracing-tester', 'tempo')
+    status = juju.wait(jubilant.all_active)
+    trace_api = status.apps['tempo-worker'].address
+
+    spans = wait_spans(trace_api, ready=lambda spans: 'ops.main' in str(spans))
+    assert 'ops.main' in [span['name'] for span in spans]
+
+    event_names = [event['name'] for event in get_events(spans)]
+    assert 'StartEvent' in event_names
+
+    checkpoint = time.time()
+    juju.run('tracing-tester/0', 'one', params={'arg': 'oneoneoneone'})
```

Comment:
I spend an embarassing amount of time trying to figure out how the action name was ending up in the span multiple times when I read the assert line haha
```py
assert 'oneoneoneone' in json.dumps(action_span)
```

Maybe a variable would be good?
```suggestion
    arg_value = 'oneoneoneoneone!!!!1'
    juju.run('tracing-tester/0', 'one', params={'arg': arg_value})
```
And later:
```py
assert arg_value in json.dumps(action_span)
```

Maybe `action_span_name = 'on action'` would be a good idea too. I'd be pretty tempted by `action_name = 'one'` and using `f'{action_name}_action'` over `'one_action'` too tbh.

Reply:
sure, going to change this up.